sympy/matrices/matrices.py

```diff
@@ -1589,6 +1583,11 @@ def _matrix_pow_by_recursion(self, num):
     def __pow__(self, num):
         if not self.is_square:
             raise NonSquareMatrixError()
+        if not sympify(num).is_integer and (num > 0) == True and not num % 1 == 0:
+            b = self.jordan_cells()[1]
```

Comment:
Computing the Jordan form is expensive. It should not be done at all when recursion is applied. Hence the role of `__pow__` is only to decide which method is to be used, Jordan forms or recursion. It can also sympify the exponent `num` so that tests like `num.is_integer` will work in `_matrix_power_by_jordan_blocks`.